#Preprocessing

##Import & Load Data

In [ ]:
#import library
import gdown
import pandas as pd
import os
import re
import nltk
from nltk.tokenize import word_tokenize
from sklearn.model_selection import train_test_split
from sklearn.metrics import mutual_info_score
import matplotlib.pyplot as plt

In [ ]:
#Connect Gdrive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
os.chdir("/content/drive/My Drive/Tugas Akhir/Dataset")
!ls

 cleaned_dataset.csv	 'Generate time.ipynb'	        result_completions.csv
 data_preprocessing.csv   hasil_cosine_similarity.csv   spreadsheet_dataset_aspek.gsheet
 dataset_aspek.xlsx	  hasil_stemming.csv	        translated.csv
 dataset_new.csv	  hasil_tokenisasi.csv	        translated.gsheet
 dataset_token.csv	  JSON
 dataset_token.gsheet	  output.json


In [ ]:
df = pd.read_excel("/content/drive/My Drive/Tugas Akhir/Dataset/dataset_aspek.xlsx")
df.head()

Unnamed: 0  No  Kalimat #                                   Review Aspek  \
0           0   1          0              pas banget rasanya di lidah  FOOD   
1           1   2          0  black peppernya yang paling enak disini  FOOD   
2           2   3          1        se’i sapi sambel lu’at nya dabest  FOOD   
3           3   4          1                   se’i sapi sambel lu’at  FOOD   
4           4   5          1                         se’i sapi sambel  FOOD   

   Sentiment  Id User   timestamp  lead time  
0          1     4810  1592928807   8.516239  
1          1     4810  1592928807   8.516239  
2          1     1541  1585775238  49.738258  
3          0     1541  1585775238  49.738258  
4          0     1541  1585775238  49.738258

##Cleansing Data

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3303 entries, 0 to 3302
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  3303 non-null   int64  
 1   No          3303 non-null   int64  
 2   Kalimat #   3303 non-null   int64  
 3   Review      3303 non-null   object 
 4   Aspek       3303 non-null   object 
 5   Sentiment   3303 non-null   int64  
 6   Id User     3303 non-null   int64  
 7   timestamp   3303 non-null   int64  
 8   lead time   3303 non-null   float64
dtypes: float64(1), int64(6), object(2)
memory usage: 232.4+ KB


In [ ]:
#mengecek data null
df.isna().sum()

Unnamed: 0    0
No            0
Kalimat #     0
Review        0
Aspek         0
Sentiment     0
Id User       0
timestamp     0
lead time     0
dtype: int64

In [ ]:
#mengecek data duplikasi
print("Jumlah Duplikasi : ", df.duplicated().sum())

Jumlah Duplikasi :  0


In [ ]:
df.describe()

Unnamed: 0           No    Kalimat #    Sentiment      Id User  \
count  3303.000000  3303.000000  3303.000000  3303.000000  3303.000000   
mean   1651.000000  1652.000000   450.418408     0.569785  2563.129882   
std     953.638296   953.638296   239.628997     0.765643  1427.931410   
min       0.000000     1.000000     0.000000    -1.000000     2.000000   
25%     825.500000   826.500000   251.000000     0.000000  1342.000000   
50%    1651.000000  1652.000000   470.000000     1.000000  2582.000000   
75%    2476.500000  2477.500000   650.500000     1.000000  3831.000000   
max    3302.000000  3303.000000   859.000000     1.000000  4998.000000   

          timestamp    lead time  
count  3.303000e+03  3303.000000  
mean   1.593650e+09    31.425311  
std    8.790629e+06    16.900461  
min    1.577880e+09     0.128850  
25%    1.585282e+09    16.848379  
50%    1.594365e+09    32.990058  
75%    1.600923e+09    45.358808  
max    1.609395e+09    59.841129

In [ ]:
df.head()

Unnamed: 0  No  Kalimat #                                   Review Aspek  \
0           0   1          0              pas banget rasanya di lidah  FOOD   
1           1   2          0  black peppernya yang paling enak disini  FOOD   
2           2   3          1        se’i sapi sambel lu’at nya dabest  FOOD   
3           3   4          1                   se’i sapi sambel lu’at  FOOD   
4           4   5          1                         se’i sapi sambel  FOOD   

   Sentiment  Id User   timestamp  lead time  
0          1     4810  1592928807   8.516239  
1          1     4810  1592928807   8.516239  
2          1     1541  1585775238  49.738258  
3          0     1541  1585775238  49.738258  
4          0     1541  1585775238  49.738258

In [ ]:
#mengecek tipe data column Review
print(df['Review'].dtype)

object


In [ ]:
#menangani invalid data pada column 'Review'
#karena tipe data dari column eview tidak valid, seharusnya string, sehingga perlu ditangani
df = df.dropna(subset=['Review'])

In [ ]:
#mengkonversi data ke string
df['Review'] = df['Review'].apply(lambda x: str(x))

##Normalisasi Data

In [ ]:
#menghilangkan simbol
def normalize_text(text):
    # Menghapus simbol atau karakter selain huruf
    cleaned_text = re.sub(r'[^a-zA-Z0-9\s]', '', text)
    return cleaned_text

df['Review'] = df['Review'].apply(normalize_text)

In [ ]:
df.head()

Unnamed: 0  No  Kalimat #                                   Review Aspek  \
0           0   1          0              pas banget rasanya di lidah  FOOD   
1           1   2          0  black peppernya yang paling enak disini  FOOD   
2           2   3          1          sei sapi sambel luat nya dabest  FOOD   
3           3   4          1                     sei sapi sambel luat  FOOD   
4           4   5          1                          sei sapi sambel  FOOD   

   Sentiment  Id User   timestamp  lead time  
0          1     4810  1592928807   8.516239  
1          1     4810  1592928807   8.516239  
2          1     1541  1585775238  49.738258  
3          0     1541  1585775238  49.738258  
4          0     1541  1585775238  49.738258

In [ ]:
df.iloc[1510:1515] #menampilkan data dari baris 1490 - 1495

Unnamed: 0    No  Kalimat #            Review    Aspek  Sentiment  \
1510        1510  1511        438  bebek goreng 810     FOOD          1   
1511        1511  1512        438  sate kambing 910     FOOD          1   
1512        1512  1513        438     sate ayam 710     FOOD          0   
1513        1513  1514        438     sekoteng 1010     FOOD          1   
1514        1514  1515        438       service 810  SERVICE          1   

      Id User   timestamp  lead time  
1510     3163  1593699825  29.701188  
1511     3163  1593699825  29.701188  
1512     3163  1593699825  29.701188  
1513     3163  1593699825  29.701188  
1514     3163  1593699825  29.701188

simbol pada kalimat nomor 2-4 sudah hilang, dan pada data nomor 1511 - 1515 simbol '/' telah hilang. Namun masih mempertahankan angka

##Case Folding
menyeragamkan semua huruf menjadi huruf kecil

In [ ]:
#mengonversi semua huruf dalam teks menjadi huruf kecil
def case_folding(text):
    # Mengonversi teks ke huruf kecil (lowercase)
    folded_text = text.lower()
    return folded_text

df['Review'] = df['Review'].apply(case_folding)

In [ ]:
df.head()

Unnamed: 0  No  Kalimat #                                   Review Aspek  \
0           0   1          0              pas banget rasanya di lidah  FOOD   
1           1   2          0  black peppernya yang paling enak disini  FOOD   
2           2   3          1          sei sapi sambel luat nya dabest  FOOD   
3           3   4          1                     sei sapi sambel luat  FOOD   
4           4   5          1                          sei sapi sambel  FOOD   

   Sentiment  Id User   timestamp  lead time  
0          1     4810  1592928807   8.516239  
1          1     4810  1592928807   8.516239  
2          1     1541  1585775238  49.738258  
3          0     1541  1585775238  49.738258  
4          0     1541  1585775238  49.738258

##Translate

In [ ]:
!pip install googletrans==4.0.0-rc1

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.1/55.1 kB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 29.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.4/133.4 kB 16.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 7.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 kB 5.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.6/53.6 kB 6.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.0/65.0 kB 8.8 MB/s eta 0:00:00
  Created wheel for googletrans: filename=googletrans-4.0.0rc1-py3-none-any.whl size=17396 sha256=2f374ed7af7e0554378b4bdd4101e4321a5cfe5cebd7f8c5ebd3f896c9d5d23e
  Stored in directory: /root/.cache/pip/wheels/c0/59/9f/7372f0cf70160fe61b528532e1a7c8498c4becd6bcffb022de
Successfully built googletrans
  Attempting uninstall: chardet
    Found existing installation: chardet 5.2.0
    Uninstalling 

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3303 entries, 0 to 3302
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  3303 non-null   int64  
 1   No          3303 non-null   int64  
 2   Kalimat #   3303 non-null   int64  
 3   Review      3303 non-null   object 
 4   Aspek       3303 non-null   object 
 5   Sentiment   3303 non-null   int64  
 6   Id User     3303 non-null   int64  
 7   timestamp   3303 non-null   int64  
 8   lead time   3303 non-null   float64
dtypes: float64(1), int64(6), object(2)
memory usage: 232.4+ KB


In [ ]:
df = df.dropna(subset=['Review'])

In [ ]:
# !pip install google-cloud-translate

In [ ]:
# from google.cloud import translate_v2 as translate
# project_id = '297062577802'
# api_key = 'AIzaSyCz6u4tO9nElC4JjF1jKPBllW21p8spe84'

# client = translate.Client.from_service_account_json(json_credentials_path='/content/service_account.json')

In [ ]:
# def translate_text(text, target_language='id'):
#     translation = client.translate(text, target_language=target_language)
#     return translation['translatedText']

In [ ]:
# df['Translated_Review'] = df['Review'].apply(translate_text)

In [ ]:
# df

In [ ]:
# csv_path = '/content/translated.csv'
# df.to_csv(csv_path, index=False)

In [ ]:
df_translated = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Dataset/translated.csv')
df_translated

Unnamed: 0    No  Kalimat #  \
0              0     1          0   
1              1     2          0   
2              2     3          1   
3              3     4          1   
4              4     5          1   
...          ...   ...        ...   
3298        3298  3299        858   
3299        3299  3300        858   
3300        3300  3301        859   
3301        3301  3302        859   
3302        3302  3303        859   

                                                 Review          Aspek  \
0                           pas banget rasanya di lidah           FOOD   
1               black peppernya yang paling enak disini           FOOD   
2                       sei sapi sambel luat nya dabest           FOOD   
3                                  sei sapi sambel luat           FOOD   
4                                       sei sapi sambel           FOOD   
...                                                 ...            ...   
3298  mushroom sauce taste just like cream sauce no ...           FOOD   
3299                                  recommended place  MISCELLANEOUS   
3300                       the dishes tasted so so good           FOOD   
3301                                    worth the price          PRICE   
3302                         the portions way too small           FOOD   

      Sentiment                                 Translated_Review  Id User  \
0             1                       pas banget rasanya di lidah     3056   
1             1           black peppernya yang paling enak disini     3056   
2             1                sei sapi sambel diambil nya dabest      833   
3             0                           sei sapi sambel diambil      833   
4             0                                   sei sapi sambel      833   
...         ...                                               ...      ...   
3298         -1  saus jamur rasanya seperti saus krim tanpa jamur     1420   
3299          1                      tempat yang direkomendasikan     1420   
3300          1                     masakannya terasa sangat enak      785   
3301          1                           sepadan dengan harganya      785   
3302         -1                            porsinya terlalu kecil      785   

       timestamp  lead time  FOOD  SERVICE  MISCELLANEOUS  AMBIENCE  PRICE  
0     1600968490  37.378804     1        0              0         0      0  
1     1600968490  37.378804     1        0              0         0      0  
2     1586127504  59.219170     1        0              0         0      0  
3     1586127504  59.219170     1        0              0         0      0  
4     1586127504  59.219170     1        0              0         0      0  
...          ...        ...   ...      ...            ...       ...    ...  
3298  1587874000  34.281820     1        0              0         0      0  
3299  1587874000  34.281820     0        0              1         0      0  
3300  1578249592  29.656725     1        0              0         0      0  
3301  1578249592  29.656725     0        0              0         0      1  
3302  1578249592  29.656725     1        0              0         0      0  

[3303 rows x 15 columns]

##Stopword Removal
Ini adalah teknik menghilangkan kata-kata yang sering digunakan yang tidak berarti dan tidak berguna untuk klasifikasi teks

In [ ]:
# import nltk
# from nltk.corpus import stopwords
# nltk.download('stopwords')

In [ ]:
# # Menginisialisasi daftar stopwords dalam bahasa Indonesia
# stop_words = set(stopwords.words('indonesian'))

In [ ]:
# #fungsi untuk stopword removal pada teks
# def remove_stopwords(text):
#     words = text.split()
#     filtered_words = [word for word in words if word.lower() not in stop_words]
#     filtered_text = ' '.join(filtered_words)
#     return filtered_text

In [ ]:
# df_translated = pd.read_csv("/content/drive/My Drive/Tugas Akhir/Dataset/translated.csv")
# df_translated.head()

In [ ]:
# # Melakukan stopword removal pada data translated
# df_translated['Cleaned_Review'] = df_translated['Translated_Review'].apply(remove_stopwords)

In [ ]:
# df_translated.to_csv('cleaned_dataset.csv', index=False)

##Stemming

Menggunakan library Sastrawi



In [ ]:
!pip install Sastrawi

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 209.7/209.7 kB 4.4 MB/s eta 0:00:00


In [ ]:
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

# Inisialisasi stemmer Sastrawi
factory = StemmerFactory()
stemmer = factory.create_stemmer()

# Contoh
text = "pas banget rasanya di lidah"

# Stemming
stemmed_text = stemmer.stem(text)

# Menampilkan teks setelah proses stemming
print(stemmed_text)

pas banget rasa di lidah


In [ ]:
df_stemming = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Dataset/translated.csv')

In [ ]:
def stemming_text(text):
  return stemmer.stem(text)

In [ ]:
def conv_to_str(text):
  return str(text)

In [ ]:
df_stemming.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3303 entries, 0 to 3302
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         3303 non-null   int64  
 1   No                 3303 non-null   int64  
 2   Kalimat #          3303 non-null   int64  
 3   Review             3303 non-null   object 
 4   Aspek              3303 non-null   object 
 5   Sentiment          3303 non-null   int64  
 6   Translated_Review  3303 non-null   object 
 7   Id User            3303 non-null   int64  
 8   timestamp          3303 non-null   int64  
 9   lead time          3303 non-null   float64
 10  FOOD               3303 non-null   int64  
 11  SERVICE            3303 non-null   int64  
 12  MISCELLANEOUS      3303 non-null   int64  
 13  AMBIENCE           3303 non-null   int64  
 14  PRICE              3303 non-null   int64  
dtypes: float64(1), int64(11), object(3)
memory usage: 387.2+ KB


In [ ]:
# df_stemming['Stemmed'] = df_stemming['Translated_Review'].apply(stemming_text)

In [ ]:
# df_stemming

In [ ]:
# Menyimpan DataFrame ke dalam file CSV
# df_stemming.to_csv('hasil_stemming.csv', index=False)

## Tokenization

In [ ]:
df_tokenize = pd.read_csv('/content/drive/MyDrive/Tugas Akhir/Dataset/hasil_stemming.csv')

In [ ]:
#download module library nltk
nltk.download('punkt')

#mengambil data hasil stemming untuk tokenisasi
text_data = df_tokenize['Stemmed']

#fungsi untuk tokenisasi
def tokenize_text(text):
    return word_tokenize(text)

# Melakukan tokenisasi pada setiap baris teks
tokenized_data = text_data.apply(tokenize_text)

# Menampilkan contoh hasil tokenisasi
print(tokenized_data.head())

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


0                  [pas, banget, rasa, di, lidah]
1    [black, peppernya, yang, paling, enak, sini]
2         [sei, sapi, sambel, ambil, nya, dabest]
3                      [sei, sapi, sambel, ambil]
4                             [sei, sapi, sambel]
Name: Stemmed, dtype: object


In [ ]:
df_tokenize['token_data'] = tokenized_data
df_tokenize

Unnamed: 0    No  Kalimat #  \
0              0     1          0   
1              1     2          0   
2              2     3          1   
3              3     4          1   
4              4     5          1   
...          ...   ...        ...   
3298        3298  3299        858   
3299        3299  3300        858   
3300        3300  3301        859   
3301        3301  3302        859   
3302        3302  3303        859   

                                                 Review          Aspek  \
0                           pas banget rasanya di lidah           FOOD   
1               black peppernya yang paling enak disini           FOOD   
2                       sei sapi sambel luat nya dabest           FOOD   
3                                  sei sapi sambel luat           FOOD   
4                                       sei sapi sambel           FOOD   
...                                                 ...            ...   
3298  mushroom sauce taste just like cream sauce no ...           FOOD   
3299                                  recommended place  MISCELLANEOUS   
3300                       the dishes tasted so so good           FOOD   
3301                                    worth the price          PRICE   
3302                         the portions way too small           FOOD   

      Sentiment                                 Translated_Review  Id User  \
0             1                       pas banget rasanya di lidah     3056   
1             1           black peppernya yang paling enak disini     3056   
2             1                sei sapi sambel diambil nya dabest      833   
3             0                           sei sapi sambel diambil      833   
4             0                                   sei sapi sambel      833   
...         ...                                               ...      ...   
3298         -1  saus jamur rasanya seperti saus krim tanpa jamur     1420   
3299          1                      tempat yang direkomendasikan     1420   
3300          1                     masakannya terasa sangat enak      785   
3301          1                           sepadan dengan harganya      785   
3302         -1                            porsinya terlalu kecil      785   

       timestamp  lead time  FOOD  SERVICE  MISCELLANEOUS  AMBIENCE  PRICE  \
0     1600968490  37.378804     1        0              0         0      0   
1     1600968490  37.378804     1        0              0         0      0   
2     1586127504  59.219170     1        0              0         0      0   
3     1586127504  59.219170     1        0              0         0      0   
4     1586127504  59.219170     1        0              0         0      0   
...          ...        ...   ...      ...            ...       ...    ...   
3298  1587874000  34.281820     1        0              0         0      0   
3299  1587874000  34.281820     0        0              1         0      0   
3300  1578249592  29.656725     1        0              0         0      0   
3301  1578249592  29.656725     0        0              0         0      1   
3302  1578249592  29.656725     1        0              0         0      0   

                                            Stemmed  \
0                          pas banget rasa di lidah   
1             black peppernya yang paling enak sini   
2                  sei sapi sambel ambil nya dabest   
3                             sei sapi sambel ambil   
4                                   sei sapi sambel   
...                                             ...   
3298  saus jamur rasa seperti saus krim tanpa jamur   
3299                        tempat yang rekomendasi   
3300                        masakan asa sangat enak   
3301                             padan dengan harga   
3302                            porsi terlalu kecil   

                                             token_data  
0                        [pas, banget, rasa, di, lidah]  
1          [black, pe

In [ ]:
# Menyimpan DataFrame ke dalam file CSV
df_tokenize.to_csv('hasil_tokenisasi.csv', index=False)

##Resample Sentiment

In [ ]:
data = df_tokenize

In [ ]:
data

Unnamed: 0    No  Kalimat #  \
0              0     1          0   
1              1     2          0   
2              2     3          1   
3              3     4          1   
4              4     5          1   
...          ...   ...        ...   
3298        3298  3299        858   
3299        3299  3300        858   
3300        3300  3301        859   
3301        3301  3302        859   
3302        3302  3303        859   

                                                 Review          Aspek  \
0                           pas banget rasanya di lidah           FOOD   
1               black peppernya yang paling enak disini           FOOD   
2                       sei sapi sambel luat nya dabest           FOOD   
3                                  sei sapi sambel luat           FOOD   
4                                       sei sapi sambel           FOOD   
...                                                 ...            ...   
3298  mushroom sauce taste just like cream sauce no ...           FOOD   
3299                                  recommended place  MISCELLANEOUS   
3300                       the dishes tasted so so good           FOOD   
3301                                    worth the price          PRICE   
3302                         the portions way too small           FOOD   

      Sentiment                                 Translated_Review  Id User  \
0             1                       pas banget rasanya di lidah     3056   
1             1           black peppernya yang paling enak disini     3056   
2             1                sei sapi sambel diambil nya dabest      833   
3             0                           sei sapi sambel diambil      833   
4             0                                   sei sapi sambel      833   
...         ...                                               ...      ...   
3298         -1  saus jamur rasanya seperti saus krim tanpa jamur     1420   
3299          1                      tempat yang direkomendasikan     1420   
3300          1                     masakannya terasa sangat enak      785   
3301          1                           sepadan dengan harganya      785   
3302         -1                            porsinya terlalu kecil      785   

       timestamp  lead time  FOOD  SERVICE  MISCELLANEOUS  AMBIENCE  PRICE  \
0     1600968490  37.378804     1        0              0         0      0   
1     1600968490  37.378804     1        0              0         0      0   
2     1586127504  59.219170     1        0              0         0      0   
3     1586127504  59.219170     1        0              0         0      0   
4     1586127504  59.219170     1        0              0         0      0   
...          ...        ...   ...      ...            ...       ...    ...   
3298  1587874000  34.281820     1        0              0         0      0   
3299  1587874000  34.281820     0        0              1         0      0   
3300  1578249592  29.656725     1        0              0         0      0   
3301  1578249592  29.656725     0        0              0         0      1   
3302  1578249592  29.656725     1        0              0         0      0   

                                            Stemmed  \
0                          pas banget rasa di lidah   
1             black peppernya yang paling enak sini   
2                  sei sapi sambel ambil nya dabest   
3                             sei sapi sambel ambil   
4                                   sei sapi sambel   
...                                             ...   
3298  saus jamur rasa seperti saus krim tanpa jamur   
3299                        tempat yang rekomendasi   
3300                        masakan asa sangat enak   
3301                             padan dengan harga   
3302                            porsi terlalu kecil   

                                             token_data  
0                        [pas, banget, rasa, di, lidah]  
1          [black, pe

In [ ]:
data['Sentiment'] = data['Sentiment'].replace(-1,0)

In [ ]:
data

Unnamed: 0    No  Kalimat #  \
0              0     1          0   
1              1     2          0   
2              2     3          1   
3              3     4          1   
4              4     5          1   
...          ...   ...        ...   
3298        3298  3299        858   
3299        3299  3300        858   
3300        3300  3301        859   
3301        3301  3302        859   
3302        3302  3303        859   

                                                 Review          Aspek  \
0                           pas banget rasanya di lidah           FOOD   
1               black peppernya yang paling enak disini           FOOD   
2                       sei sapi sambel luat nya dabest           FOOD   
3                                  sei sapi sambel luat           FOOD   
4                                       sei sapi sambel           FOOD   
...                                                 ...            ...   
3298  mushroom sauce taste just like cream sauce no ...           FOOD   
3299                                  recommended place  MISCELLANEOUS   
3300                       the dishes tasted so so good           FOOD   
3301                                    worth the price          PRICE   
3302                         the portions way too small           FOOD   

      Sentiment                                 Translated_Review  Id User  \
0             1                       pas banget rasanya di lidah     3056   
1             1           black peppernya yang paling enak disini     3056   
2             1                sei sapi sambel diambil nya dabest      833   
3             0                           sei sapi sambel diambil      833   
4             0                                   sei sapi sambel      833   
...         ...                                               ...      ...   
3298          0  saus jamur rasanya seperti saus krim tanpa jamur     1420   
3299          1                      tempat yang direkomendasikan     1420   
3300          1                     masakannya terasa sangat enak      785   
3301          1                           sepadan dengan harganya      785   
3302          0                            porsinya terlalu kecil      785   

       timestamp  lead time  FOOD  SERVICE  MISCELLANEOUS  AMBIENCE  PRICE  \
0     1600968490  37.378804     1        0              0         0      0   
1     1600968490  37.378804     1        0              0         0      0   
2     1586127504  59.219170     1        0              0         0      0   
3     1586127504  59.219170     1        0              0         0      0   
4     1586127504  59.219170     1        0              0         0      0   
...          ...        ...   ...      ...            ...       ...    ...   
3298  1587874000  34.281820     1        0              0         0      0   
3299  1587874000  34.281820     0        0              1         0      0   
3300  1578249592  29.656725     1        0              0         0      0   
3301  1578249592  29.656725     0        0              0         0      1   
3302  1578249592  29.656725     1        0              0         0      0   

                                            Stemmed  \
0                          pas banget rasa di lidah   
1             black peppernya yang paling enak sini   
2                  sei sapi sambel ambil nya dabest   
3                             sei sapi sambel ambil   
4                                   sei sapi sambel   
...                                             ...   
3298  saus jamur rasa seperti saus krim tanpa jamur   
3299                        tempat yang rekomendasi   
3300                        masakan asa sangat enak   
3301                             padan dengan harga   
3302                            porsi terlalu kecil   

                                             token_data  
0                        [pas, banget, rasa, di, lidah]  
1          [black, pe

In [ ]:
import seaborn as sns
import plotly.express as px
fig = px.bar(data.groupby('Sentiment').size().reset_index(name='count'),
             x='Sentiment', y='count', width=480, height=480,
             barmode='group', title='Distribusi label')
fig.show(renderer='colab')

In [ ]:
#Aspect Distribution
sorted_data = data.groupby('Aspek').size().reset_index(name='count').sort_values(by='count', ascending=False)

fig = px.bar(data.groupby('Aspek').size().reset_index(name='count'),
             x='Aspek', y='count', width=480, height=480,
             barmode='group', title='Distribusi berdasarkan Aspek', category_orders={"Aspek": sorted_data["Aspek"].tolist()})
fig.show(renderer='colab')

In [ ]:
# Menyimpan DataFrame ke dalam file CSV
data.to_csv('data_preprocessing.csv', index=False)

## Convert Text List

In [ ]:
import ast

In [ ]:
#masukkan kata ke dalam list
def convert_text_list(texts):
    # texts = ast.literal_eval(texts)
    # return [text for text in texts]
    temp = []
    for i in texts.split():
      temp.append([i])
    return temp
def listToString(s):

    # initialize an empty string
    str1 = " "

    # return string
    return (str1.join(s))

In [ ]:
data['rev_list'] = data['Stemmed'].apply(convert_text_list) #with stemming
# data['rev_kalimat'] = data['rev_list'].apply(listToString)

In [ ]:
data

Unnamed: 0    No  Kalimat #  \
0              0     1          0   
1              1     2          0   
2              2     3          1   
3              3     4          1   
4              4     5          1   
...          ...   ...        ...   
3298        3298  3299        858   
3299        3299  3300        858   
3300        3300  3301        859   
3301        3301  3302        859   
3302        3302  3303        859   

                                                 Review          Aspek  \
0                           pas banget rasanya di lidah           FOOD   
1               black peppernya yang paling enak disini           FOOD   
2                       sei sapi sambel luat nya dabest           FOOD   
3                                  sei sapi sambel luat           FOOD   
4                                       sei sapi sambel           FOOD   
...                                                 ...            ...   
3298  mushroom sauce taste just like cream sauce no ...           FOOD   
3299                                  recommended place  MISCELLANEOUS   
3300                       the dishes tasted so so good           FOOD   
3301                                    worth the price          PRICE   
3302                         the portions way too small           FOOD   

      Sentiment                                 Translated_Review  Id User  \
0             1                       pas banget rasanya di lidah     3056   
1             1           black peppernya yang paling enak disini     3056   
2             1                sei sapi sambel diambil nya dabest      833   
3             0                           sei sapi sambel diambil      833   
4             0                                   sei sapi sambel      833   
...         ...                                               ...      ...   
3298          0  saus jamur rasanya seperti saus krim tanpa jamur     1420   
3299          1                      tempat yang direkomendasikan     1420   
3300          1                     masakannya terasa sangat enak      785   
3301          1                           sepadan dengan harganya      785   
3302          0                            porsinya terlalu kecil      785   

       timestamp  lead time  FOOD  SERVICE  MISCELLANEOUS  AMBIENCE  PRICE  \
0     1600968490  37.378804     1        0              0         0      0   
1     1600968490  37.378804     1        0              0         0      0   
2     1586127504  59.219170     1        0              0         0      0   
3     1586127504  59.219170     1        0              0         0      0   
4     1586127504  59.219170     1        0              0         0      0   
...          ...        ...   ...      ...            ...       ...    ...   
3298  1587874000  34.281820     1        0              0         0      0   
3299  1587874000  34.281820     0        0              1         0      0   
3300  1578249592  29.656725     1        0              0         0      0   
3301  1578249592  29.656725     0        0              0         0      1   
3302  1578249592  29.656725     1        0              0         0      0   

                                            Stemmed  \
0                          pas banget rasa di lidah   
1             black peppernya yang paling enak sini   
2                  sei sapi sambel ambil nya dabest   
3                             sei sapi sambel ambil   
4                                   sei sapi sambel   
...                                             ...   
3298  saus jamur rasa seperti saus krim tanpa jamur   
3299                        tempat yang rekomendasi   
3300                        masakan asa sangat enak   
3301                             padan dengan harga   
3302                            porsi terlalu kecil   

                                             token_data  \
0                        [pas, banget, rasa, di, lidah]   
1          [black, 

##Data JSON

In [ ]:
# sentiment_result = pd.read_json('/content/drive/MyDrive/Tugas Akhir/result.json')

In [ ]:
# json = sentiment_result['completions'].to_json(orient='records', lines=True)

In [ ]:
# with open('output.json', 'w') as json_file:
#   json_file.write(json)

In [ ]:
# sentiment_result = pd.read_json('/content/drive/MyDrive/Tugas Akhir/Dataset/output.json', lines=True)

In [ ]:
# dict_sr = sentiment_result.to_dict(orient='records')

In [ ]:
# df_sr = pd.DataFrame()
# for i in dict_sr:
#   if not not i.get(0).get('result'):
#     new_row = pd.DataFrame(
#         {'Text': [i.get(0).get('result')[0].get('value').get('text')],
#          'Label': [i.get(0).get('result')[0].get('value').get('labels')[1]],
#          'Sentiment': [i.get(0).get('result')[0].get('value').get('labels')[0]]}
#         )
#     df_sr = pd.concat([df_sr, new_row], ignore_index=True)

In [ ]:
# df_sr

In [ ]:
# df_sr.to_csv('dataset_new.csv', index=False)

# Oversampling

In [ ]:
from imblearn.over_sampling import RandomOverSampler
cc = RandomOverSampler()

In [ ]:
X_sample = data[['Stemmed', 'FOOD', 'SERVICE', 'MISCELLANEOUS', 'AMBIENCE', 'PRICE']]
Y_sample = data['Sentiment']

In [ ]:
X_resampled, Y_resampled = cc.fit_resample(X_sample, Y_sample)

In [ ]:
df_resampled = pd.DataFrame(X_resampled)
df_resampled = df_resampled.join(Y_resampled)

In [ ]:
import seaborn as sns
import plotly.express as px
fig = px.bar(df_resampled.groupby('Sentiment').size().reset_index(name='count'),
             x='Sentiment', y='count', width=480, height=480,
             barmode='group', title='Sentiment Distribution')
fig.show(renderer='colab')

#Split Data

In [ ]:
# #Split Data
# df_x = df_tokenize.drop(labels=['Sentiment','#'], axis=1)
# df_y = df_tokenize["Sentiment"]
# X_train,X_test,Y_train,Y_test=train_test_split(df_x,df_y,test_size=0.75)
# print("jumlah data train :",len(X_train))
# print("jumlah data tes :",len(X_test))

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df_resampled[['Aspek', 'Stemmed', 'token_data', 'rev_list', 'Id User', 'timestamp', 'lead time', 'FOOD', 'SERVICE', 'MISCELLANEOUS', 'AMBIENCE', 'PRICE', ]], df_resampled['Sentiment'], test_size = 0.3, random_state= 42, shuffle = True)

data_train = pd.DataFrame(y_train)
data_train = data_train.join(X_train)

data_test = pd.DataFrame(y_test)
data_test = data_test.join(X_test)

KeyError: ignored

In [ ]:
data_train

In [ ]:
data_test

#Ekstraksi Fitur

Menggunakan Word2Vec (Skip-Gram) untuk menghasilkan representasi vektor kata-kata berdasarkan konteks kata tersebut dalam teks.

In [ ]:
from gensim.models import Word2Vec
from gensim.models import KeyedVectors
from gensim.models.phrases import Phrases, Phraser
from nltk.tokenize import word_tokenize
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np
import pickle as pkl

In [ ]:
data_train

In [ ]:
model = KeyedVectors.load_word2vec_format('/content/drive/MyDrive/Tugas Akhir/word2vec/ws_model100.bin',binary=True)

In [ ]:
# Split data into training and testing sets before Word2Vec
X = np.array(df_resampled[['Stemmed', 'FOOD', 'SERVICE', 'MISCELLANEOUS', 'AMBIENCE', 'PRICE']])
y = np.array(df_resampled['Sentiment'])

# Split data for Word2Vec training (8746 data) and SVM training (1000 data)
X_w2v_train, X_svm, y_w2v_train, y_svm = train_test_split(X, y, test_size=1000, random_state=42)

In [ ]:
aspect_w2v_train = np.array(X_w2v_train[:, 1:6], dtype='int')

In [ ]:
X_train_tokenized = [tokenize_text(text) for text in X_w2v_train[:, 0]]

In [ ]:
def get_w2v_features(w2v_model, sentence_group):
    feature_vec = np.zeros(w2v_model.vector_size, dtype="float32")
    num_words = 0

    for word in sentence_group:
        if word in w2v_model:
            feature_vec = np.add(feature_vec, w2v_model[word])
            num_words += 1

    if num_words > 0:
        feature_vec = np.divide(feature_vec, num_words)

    return feature_vec

In [ ]:
# Calculate Word2Vec features for training data
X_train_w2v = [get_w2v_features(model, tokens) for tokens in X_train_tokenized]

In [ ]:
#membuat matrix untuk array multidimensi
matrix_w2v = np.array(X_train_w2v)

In [ ]:
matrix_w2v.shape

In [ ]:
aspect_w2v_train.shape

In [ ]:
np.array(aspect_w2v_train[:,0]).shape

In [ ]:
matrix_w2v

In [ ]:
# Menambahkan feature aspek kembali
# matrix_w2v = np.hstack((matrix_w2v, np.array(aspect_w2v_train)))

In [ ]:
matrix_w2v.shape

Test data

In [ ]:
# X_svm[:, 0]

In [ ]:
aspect_w2v_test = np.array(X_svm[:, 1:6], dtype='int')

In [ ]:
X_test_tokenized = [tokenize_text(text) for text in X_svm[:, 0]]

In [ ]:
# Calculate Word2Vec features for testing data
X_test_w2v = [get_w2v_features(model, tokens) for tokens in X_test_tokenized]

In [ ]:
#membuat matrix untuk array multidimensi
matrix_w2v_test = np.array(X_test_w2v)

In [ ]:
# matrix_w2v_test = np.hstack((matrix_w2v_test, np.array(aspect_w2v_test)))

In [ ]:
# data['rev_list']

In [ ]:
# def get_vector(text):
#   vec = []
#   for i in text:
#     for o in i:
#       try:
#         vec.append(model[o])
#       except:
#         print('', end='')
#   return np.asarray(vec).astype('float')

Train Data

In [ ]:
# data_train['vectorized'] = data_train['rev_list'].apply(get_vector)

In [ ]:
# data_train

Test Data

In [ ]:
# data_test['vectorized'] = data_test['rev_list'].apply(get_vector)

In [ ]:
# data_test['vectorized']

#Seleksi Fitur (CFS)
CFS untuk menghitung bobot setiap fitur dalam dataset. Kemudian dilakukan pemilihan karakteristik dengan menggunakan kriteria pembobotan yang diurutkan dari besar ke kecil.

In [ ]:
# np.corrcoef(matrix_w2v).shape

In [ ]:
# 1. Calculate the correlation coefficients
correlation_matrix = np.corrcoef(matrix_w2v, rowvar=False)

# 2. Identify columns with correlation coefficient
correlation_threshold = 0.1
columns_to_drop = []
for i in range(correlation_matrix.shape[1]):
    for j in range(i + 1, correlation_matrix.shape[1]):
        if abs(correlation_matrix[i, j]) < correlation_threshold :
            columns_to_drop.append(j)

# 3. Remove columns with low correlation
data_filtered_train = np.delete(matrix_w2v, columns_to_drop, axis=1)
data_filtered_test = np.delete(matrix_w2v_test, columns_to_drop, axis=1)
print("Original Data:\n", matrix_w2v)
print("Data after dropping low-correlation columns train data:\n", data_filtered_train)
print("Data after dropping low-correlation columns test data:\n", data_filtered_test)

In [ ]:
data_filtered_train = np.hstack((data_filtered_train, aspect_w2v_train))

In [ ]:
data_filtered_train

In [ ]:
data_filtered_test = np.hstack((data_filtered_test, aspect_w2v_test))

# SVM

In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
# Inisialisasi model SVM dengan CFS
svm_model = SVC(kernel='rbf')  # Anda bisa memilih kernel yang sesuai (linear, rbf, dll.)

# Latih model SVM pada data latih
svm_model.fit(data_filtered_train, y_w2v_train)

# Lakukan prediksi pada data uji
y_pred = svm_model.predict(data_filtered_test)

# Evaluasi model
report = classification_report(y_pred, y_svm, digits=5)
print(report)

In [ ]:
# Inisialisasi model SVM Without CFS
svm_model = SVC(kernel='rbf')  # Anda bisa memilih kernel yang sesuai (linear, rbf, dll.)

# Latih model SVM pada data latih
svm_model.fit(matrix_w2v, y_w2v_train)

# Lakukan prediksi pada data uji
y_pred = svm_model.predict(matrix_w2v_test)

# Evaluasi model
report = classification_report(y_pred, y_svm, digits=5)
print(report)

In [ ]:
from sklearn.model_selection import cross_val_score, KFold

In [ ]:
aspect_data = np.array(X[:, 1:6], dtype='int')

In [ ]:
aspect_data

In [ ]:
X_tokenized = [tokenize_text(text) for text in X[:, 0]]

In [ ]:
X_w2v = [get_w2v_features(model, tokens) for tokens in X_tokenized]

In [ ]:
kf_filtered = np.delete(X_w2v, columns_to_drop, axis=1)

In [ ]:
kf_filtered = np.hstack((kf_filtered, aspect_data))

In [ ]:
X_w2v = np.hstack((X_w2v, aspect_data))

In [ ]:
kf_filtered = np.hstack((kf_filtered, aspect_data))

In [ ]:
X_w2v

In [ ]:
# Inisialisasi model SVM
svm_model_kf_normal = SVC(kernel='rbf')  # Anda bisa memilih kernel yang sesuai (linear, rbf, dll.)

# Latih model SVM pada data latih
svm_model_kf_normal.fit(X_w2v, y)

In [ ]:
# Inisialisasi model SVM
svm_model_kf_filtered = SVC(kernel='rbf')  # Anda bisa memilih kernel yang sesuai (linear, rbf, dll.)

# Latih model SVM pada data latih
svm_model_kf_filtered.fit(kf_filtered, y)

In [ ]:
# Cross Validation untuk data tanpa CFS
kfold = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(svm_model_kf_normal, X_w2v, y, cv=kfold)
print("Akurasi Rata-rata K-Fold Cross-Validation:", scores.mean())

In [ ]:
 # Cross Validation untuk data dengan CFS
kfold = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(svm_model_kf_filtered, kf_filtered, y, cv=kfold)
print("Akurasi Rata-rata K-Fold Cross-Validation:", scores.mean())